pkg/iobuf/bufferedreaderseeker.go

```diff
@@ -0,0 +1,198 @@
+package iobuf
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+)
+
+// BufferedReadSeeker provides a buffered reading interface with seeking capabilities.
+// It wraps an io.Reader and optionally an io.Seeker, allowing for efficient
+// reading and seeking operations, even on non-seekable underlying readers.
+type BufferedReadSeeker struct {
+	reader io.Reader
+	seeker io.Seeker // If the reader supports seeking, it's stored here for direct access
+
+	buffer *bytes.Buffer // Internal buffer to store read bytes for non-seekable readers
+
+	bytesRead int64 // Total number of bytes read from the underlying reader
+	index     int64 // Current position in the virtual stream
+
+	// Flag to control buffering. This flag is used to indicate whether buffering is active.
+	// Buffering is enabled during initial reads (e.g., for MIME type detection and format identification).
+	// Once these operations are done, buffering can be disabled to prevent further writes to the buffer
+	// and to optimize subsequent reads directly from the underlying reader.
+	activeBuffering bool
```

Comment:
What happens if we try to seek on a non-seeker reader and active buffering is disabled?

Reply:
We might receive invalid data. We can add a check to prevent this.

Reply:
Actually, i'll have to include that logic in the next [PR](https://github.com/trufflesecurity/trufflehog/pull/3052).